crates/sol-macro/src/expand/function.rs

```diff
@@ -15,74 +16,121 @@ use syn::{Result, Token};
+fn expand_return_struct_def(cx: &ExpCtxt<'_>, function: &ItemFunction) -> TokenStream {
+    let attrs = &function.attrs;
+    let return_name = cx.return_name(function);
+    let fields = if let Some(ref returns) = function.returns {
+        expand_fields(&returns.returns).collect::<Vec<_>>()
+    } else {
+        vec![]
+    };
+
+    quote! {
+        #(#attrs)*
+        #[allow(non_camel_case_types, non_snake_case)]
+        #[derive(Clone)]
+        pub struct #return_name {
+            #(pub #fields,)*
+        }
+    }
+}
 
-    let fields = expand_fields(params);
+fn expand_call(cx: &ExpCtxt<'_>, function: &ItemFunction) -> Result<TokenStream> {
+    cx.assert_resolved(&function.arguments)?;
+    if let Some(ref returns) = function.returns {
+        cx.assert_resolved(&returns.returns)?;
+    }
 
-    let signature = cx.signature(function.name().as_string(), params);
-    let selector = crate::utils::selector(&signature);
+    let call_name = cx.call_name(function);
+    let return_name = cx.return_name(function);
 
-    let converts = expand_from_into_tuples(call_name, params);
+    let struct_def = expand_call_struct_def(cx, function);
+    let return_def = expand_return_struct_def(cx, function);
 
-    let tokenize_impl = if params.is_empty() {
-        quote! { ::core::convert::From::from([]) }
+    let call_tuple = expand_tuple_types(function.arguments.types()).0;
+    let return_tuple = if let Some(returns) = &function.returns {
+        expand_tuple_types(returns.returns.types()).0
     } else {
-        expand_tokenize_func(params.iter())
+        quote! { () }
     };
 
-    let return_tuple_type = function
+    let converts = expand_from_into_tuples(&call_name, &function.arguments);
+
+    let return_converts = function
         .returns
         .as_ref()
-        .map(|returns| {
-            let returns = &returns.returns;
-            let (fields, _) = expand_tuple_types(returns.types());
-            fields
-        })
-        .unwrap_or_else(|| quote! { () });
+        .map(|returns| expand_from_into_tuples(&return_name, &returns.returns))
+        .unwrap_or_else(|| expand_from_into_unit(&return_name));
+
+    let signature = cx.signature(function.name().as_string(), &function.arguments);
+    let selector = crate::utils::selector(&signature);
+
+    let tokenize_impl = if function.arguments.is_empty() {
+        quote! { ::core::convert::From::from([]) }
+    } else {
+        expand_tokenize_func(function.arguments.iter())
+    };
 
-    let attrs = &function.attrs;
     let tokens = quote! {
-        #(#attrs)*
-        #[allow(non_camel_case_types, non_snake_case)]
-        #[derive(Clone)]
-        pub struct #call_name {
-            #(pub #fields,)*
-        }
+        #struct_def
+        #return_def
 
         #[allow(non_camel_case_types, non_snake_case, clippy::style)]
         const _: () = {
-            #converts
+            { #converts }
+            { #return_converts }
```

Comment:
what's the block for

Reply:
both generate a type alias named UnderlyingSolTuple. the scope prevents a name conflict